### PR TITLE
Spark 3.1: Add prefix mismatch mode for deleting orphan files

### PR DESCRIPTION
**Pull request by manuzhang**
Back-port of https://github.com/apache/iceberg/pull/4652 to Spark 3.1

### CODE REVIEW

**Comment by manuzhang** [comment]
@aokolnychyi @karuppayya Please help review the back-port. I think it's critical for Spark 3.1 users as the prefix mismatched has caused failure of maintenance jobs (e.g. ExpireSnapshots) in our cluster.

**Comment by manuzhang** [comment]
Can anyone take a look at this PR?

**Review by szehon-ho — APPROVED** [review]
Formatting looks different, but logic looks good to me.

**Comment by manuzhang** [comment]
Any more comments here?

File: spark/v3.1/spark/src/main/java/org/apache/iceberg/spark/actions/BaseDeleteOrphanFilesSparkAction.java [hunk 1]
```diff
@@ -296,4 +334,160 @@ private static FlatMapFunction<Iterator<String>, String> listDirsRecursively(
...
+        }
+      }
+    }
+    return flattenedMap;
+  }
+
+  private static MapPartitionsFunction<Row, FileMetadata> toFileMetadata(
+      Map<String, String> equalSchemesMap, Map<String, String> equalAuthoritiesMap) {
+    return rows ->
+        Iterators.transform(
```
Comment (szehon-ho) [inline spark/v3.1/spark/src/main/java/org/apache/iceberg/spark/actions/BaseDeleteOrphanFilesSparkAction.java:434]: Same.

File: spark/v3.1/spark/src/main/java/org/apache/iceberg/spark/actions/BaseDeleteOrphanFilesSparkAction.java [hunk 2]
```diff
@@ -296,4 +334,160 @@ private static FlatMapFunction<Iterator<String>, String> listDirsRecursively(
...
+
+    return orphanFiles;
+  }
+
+  private static MapPartitionsFunction<Tuple2<FileMetadata, FileMetadata>, String> findOrphanFiles(
+      PrefixMismatchMode mode, SetAccumulator<Pair<String, String>> conflicts) {
+    return rows -> {
+      Iterator<String> transformed =
+          Iterators.transform(
+              rows,
```
Comment (szehon-ho) [inline spark/v3.1/spark/src/main/java/org/apache/iceberg/spark/actions/BaseDeleteOrphanFilesSparkAction.java:385]: Nit: could we quickly fix the formatting?  It should be on same line like in Spark 3.2?

`Iterator<String> transformed = Iterators.transform(rows, row -> {`

File: spark/v3.1/spark/src/main/java/org/apache/iceberg/spark/actions/BaseDeleteOrphanFilesSparkAction.java [hunk 3]
```diff
@@ -296,4 +334,160 @@ private static FlatMapFunction<Iterator<String>, String> listDirsRecursively(
...
+                  }
+                }
+
+                return null;
+              });
+      return Iterators.filter(transformed, Objects::nonNull);
+    };
+  }
+
+  private static Map<String, String> flattenMap(Map<String, String> map) {
```
Comment (szehon-ho) [inline spark/v3.1/spark/src/main/java/org/apache/iceberg/spark/actions/BaseDeleteOrphanFilesSparkAction.java:418]: Should this method be in the same place, after the first findOrphanFiles(), for easier cherry-pick of further changes?

**Review by szehon-ho — COMMENTED** [review]
Some formatting comments.

We should probably also backport https://github.com/apache/iceberg/pull/5495 as well then, as this is probably not very performant?

File: spark/v3.1/spark/src/main/java/org/apache/iceberg/spark/actions/BaseDeleteOrphanFilesSparkAction.java (hunk 2 above)
Comment (szehon-ho) [reply spark/v3.1/spark/src/main/java/org/apache/iceberg/spark/actions/BaseDeleteOrphanFilesSparkAction.java:398]: Hm, I dont how did it work in :  https://github.com/apache/iceberg/pull/4652/files   its ok then.

**Comment by szehon-ho** [comment]
Thanks @manuzhang

**Comment by manuzhang** [comment]
@szehon-ho I'm wondering we still need to backport #5495 since from [Anton's comment](https://github.com/apache/iceberg/pull/5495#issuecomment-1219689082) there's no big difference after the change.

**Pull request merged**